File: sdk/ios/Sources/MaxAuth/MaxAuth.swift
import Foundation
import CryptoKit
import Security

public final class MaxAuth {
    public struct AuthorizationRequest {
        public let authorizationURL: URL
        public let codeVerifier: String
        public let state: String
    }

    public struct TokenSet: Codable, Sendable {
        public let accessToken: String
        public let refreshToken: String?
        public let tokenType: String
        public let expiresIn: Int
        public let scope: String?
        public let idToken: String?

        public init(accessToken: String, refreshToken: String?, tokenType: String, expiresIn: Int, scope: String?, idToken: String?) {
            self.accessToken = accessToken
            self.refreshToken = refreshToken
            self.tokenType = tokenType
            self.expiresIn = expiresIn
            self.scope = scope
            self.idToken = idToken
        }
    }

    public struct UserInfo: Codable, Sendable {
        public let subject: String
        public let name: String?
        public let username: String?
        public let email: String?
        public let emailVerified: Bool?
        public let picture: String?

        enum CodingKeys: String, CodingKey {
            case subject = "sub"
            case name
            case username = "preferred_username"
            case email
            case emailVerified = "email_verified"
            case picture
        }
    }

    private let clientID: String
    private let redirectURI: String
    private let issuer: String

    public init(clientID: String, redirectURI: String, issuer: String = "https://auth.max-ai.name.ng") {
        self.clientID = clientID
        self.redirectURI = redirectURI
        self.issuer = issuer
    }

    public func createAuthorizationRequest(scopes: [String] = ["openid", "profile", "email"]) -> AuthorizationRequest {
        let verifier = Self.randomURLSafe(length: 64)
        let challenge = Self.sha256Base64URL(verifier)
        let state = Self.randomURLSafe(length: 32)
        var components = URLComponents(string: "\(issuer)/api/v1/oauth/authorize")!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: scopes.joined(separator: " ")),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        return AuthorizationRequest(authorizationURL: components.url!, codeVerifier: verifier, state: state)
    }

    public func isExpectedCallback(_ url: URL, state: String) -> Bool {
        guard let expected = URL(string: redirectURI) else { return false }
        return url.scheme == expected.scheme && url.host == expected.host && url.port == expected.port &&
            URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems?.first(where: { $0.name == "state" })?.value == state
    }

    public func authorizationCode(from callback: URL, state: String) throws -> String {
        guard isExpectedCallback(callback, state: state) else { throw MaxAuthError.invalidCallback }
        let components = URLComponents(url: callback, resolvingAgainstBaseURL: false)
        if let error = components?.queryItems?.first(where: { $0.name == "error" })?.value {
            throw MaxAuthError.authorizationFailed(error)
        }
        guard let code = components?.queryItems?.first(where: { $0.name == "code" })?.value else {
            throw MaxAuthError.missingAuthorizationCode
        }
        return code
    }

    public func exchangeCode(code: String, codeVerifier: String) async throws -> TokenSet {
        var request = URLRequest(url: URL(string: "\(issuer)/api/v1/oauth/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = form([
            ("grant_type", "authorization_code"),
            ("client_id", clientID),
            ("code", code),
            ("redirect_uri", redirectURI),
            ("code_verifier", codeVerifier)
        ])
        return try await perform(request, as: TokenSet.self)
    }

    public func refresh(refreshToken: String) async throws -> TokenSet {
        var request = URLRequest(url: URL(string: "\(issuer)/api/v1/oauth/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = form([
            ("grant_type", "refresh_token"),
            ("client_id", clientID),
            ("refresh_token", refreshToken)
        ])
        return try await perform(request, as: TokenSet.self)
    }

    public func revoke(token: String) async throws {
        var request = URLRequest(url: URL(string: "\(issuer)/api/v1/oauth/revoke")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = form([("token", token), ("client_id", clientID)])
        _ = try await URLSession.shared.data(for: request)
    }

    public func userInfo(accessToken: String) async throws -> UserInfo {
        var request = URLRequest(url: URL(string: "\(issuer)/api/v1/oauth/userinfo")!)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        return try await perform(request, as: UserInfo.self)
    }

    public func saveTokenSet(_ tokenSet: TokenSet) throws {
        let data = try JSONEncoder().encode(tokenSet)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "MAXAuth",
            kSecAttrAccount as String: clientID,
            kSecValueData as String: data
        ]
        SecItemDelete(query as CFDictionary)
        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else { throw MaxAuthError.keychain(status) }
    }

    public func loadTokenSet() throws -> TokenSet? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "MAXAuth",
            kSecAttrAccount as String: clientID,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = result as? Data else { throw MaxAuthError.keychain(status) }
        return try JSONDecoder().decode(TokenSet.self, from: data)
    }

    public func clearTokenSet() throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "MAXAuth",
            kSecAttrAccount as String: clientID
        ]
        let status = SecItemDelete(query as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else { throw MaxAuthError.keychain(status) }
    }

    private func perform<T: Decodable>(_ request: URLRequest, as type: T.Type) async throws -> T {
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw MaxAuthError.invalidResponse }
        guard (200...299).contains(http.statusCode) else {
            let message = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String
            throw MaxAuthError.http(http.statusCode, message ?? "MAX Auth request failed")
        }
        return try JSONDecoder().decode(type, from: data)
    }

    private func form(_ values: [(String, String)]) -> Data? {
        values.map { "\($0.0.urlEncoded)=\($0.1.urlEncoded)" }.joined(separator: "&").data(using: .utf8)
    }

    private static func randomURLSafe(length: Int) -> String {
        let alphabet = Array("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~")
        return String((0..<length).compactMap { _ in alphabet.randomElement() })
    }

    private static func sha256Base64URL(_ value: String) -> String {
        let digest = SHA256.hash(data: Data(value.utf8))
        return Data(digest).base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

public enum MaxAuthError: Error {
    case invalidCallback
    case authorizationFailed(String)
    case missingAuthorizationCode
    case invalidResponse
    case http(Int, String)
    case keychain(OSStatus)
}

private extension String {
    var urlEncoded: String {
        addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed.subtracting(CharacterSet(charactersIn: "&+="))) ?? self
    }
}

File: sdk/android/src/main/kotlin/ng/name/maxauth/MaxAuth.kt
package ng.name.maxauth

import android.content.Context
import android.net.Uri
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.KeyStore
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * MAX Auth client for Android public OAuth clients.
 *
 * Uses Authorization Code + S256 PKCE. Client secrets must never be shipped
 * in an Android application. Network calls should be made off the main thread.
 */
class MaxAuth(
    private val clientId: String,
    private val redirectUri: String,
    private val issuer: String = "https://auth.max-ai.name.ng"
) {
    data class AuthorizationRequest(
        val authorizationUrl: String,
        val codeVerifier: String,
        val state: String
    )

    data class TokenSet(
        val accessToken: String,
        val refreshToken: String?,
        val tokenType: String,
        val expiresIn: Long,
        val scope: String?,
        val idToken: String?
    )

    data class UserInfo(
        val subject: String,
        val name: String?,
        val username: String?,
        val email: String?,
        val emailVerified: Boolean?,
        val picture: String?
    )

    fun createAuthorizationRequest(scopes: List<String> = listOf("openid", "profile", "email")): AuthorizationRequest {
        val verifier = randomUrlSafe(64)
        val challenge = sha256Base64Url(verifier)
        val state = randomUrlSafe(32)
        val uri = Uri.parse("$issuer/api/v1/oauth/authorize").buildUpon()
            .appendQueryParameter("client_id", clientId)
            .appendQueryParameter("redirect_uri", redirectUri)
            .appendQueryParameter("response_type", "code")
            .appendQueryParameter("scope", scopes.joinToString(" "))
            .appendQueryParameter("state", state)
            .appendQueryParameter("code_challenge", challenge)
            .appendQueryParameter("code_challenge_method", "S256")
            .build()
        return AuthorizationRequest(uri.toString(), verifier, state)
    }

    fun isExpectedCallback(callbackUri: Uri, expectedState: String): Boolean {
        val expected = Uri.parse(redirectUri)
        return callbackUri.scheme == expected.scheme &&
            callbackUri.host == expected.host &&
            callbackUri.port == expected.port &&
            callbackUri.getQueryParameter("state") == expectedState
    }

    fun authorizationCode(callbackUri: Uri, expectedState: String): String {
        if (!isExpectedCallback(callbackUri, expectedState)) {
            throw MaxAuthException("Invalid OAuth callback state or redirect URI")
        }
        callbackUri.getQueryParameter("error")?.let { error ->
            throw MaxAuthException("MAX Auth authorization failed: $error")
        }
        return callbackUri.getQueryParameter("code")
            ?: throw MaxAuthException("MAX Auth callback did not contain an authorization code")
    }

    fun exchangeCode(code: String, codeVerifier: String): TokenSet {
        val body = form(
            "grant_type" to "authorization_code",
            "client_id" to clientId,
            "code" to code,
            "redirect_uri" to redirectUri,
            "code_verifier" to codeVerifier
        )
        return parseTokenSet(post("$issuer/api/v1/oauth/token", body))
    }

    fun refresh(refreshToken: String): TokenSet {
        val body = form(
            "grant_type" to "refresh_token",
            "client_id" to clientId,
            "refresh_token" to refreshToken
        )
        return parseTokenSet(post("$issuer/api/v1/oauth/token", body))
    }

    fun revoke(token: String) {
        post(
            "$issuer/api/v1/oauth/revoke",
            form("token" to token, "client_id" to clientId)
        )
    }

    fun userInfo(accessToken: String): UserInfo {
        val response = request("GET", "$issuer/api/v1/oauth/userinfo", null, accessToken)
        val json = JSONObject(response)
        return UserInfo(
            subject = json.getString("sub"),
            name = json.optStringOrNull("name"),
            username = json.optStringOrNull("preferred_username"),
            email = json.optStringOrNull("email"),
            emailVerified = if (json.has("email_verified")) json.optBoolean("email_verified") else null,
            picture = json.optStringOrNull("picture")
        )
    }

    fun saveTokenSet(context: Context, tokenSet: TokenSet) {
        SecureTokenStore(context, clientId).save(tokenSet)
    }

    fun loadTokenSet(context: Context): TokenSet? = SecureTokenStore(context, clientId).load()

    fun clearTokenSet(context: Context) = SecureTokenStore(context, clientId).clear()

    private fun post(url: String, body: String): String = request("POST", url, body, null)

    private fun request(method: String, url: String, body: String?, bearer: String?): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 20_000
            setRequestProperty("Accept", "application/json")
            if (bearer != null) setRequestProperty("Authorization", "Bearer $bearer")
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            }
        }
        try {
            if (body != null) connection.outputStream.use { it.write(body.toByteArray(StandardCharsets.UTF_8)) }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val response = BufferedReader(InputStreamReader(stream, StandardCharsets.UTF_8)).use { it.readText() }
            if (status !in 200..299) {
                val message = runCatching { JSONObject(response).optString("message") }.getOrNull()
                throw MaxAuthException(message?.takeIf { it.isNotBlank() } ?: "MAX Auth request failed ($status)")
            }
            return response
        } finally {
            connection.disconnect()
        }
    }

    private fun parseTokenSet(response: String): TokenSet {
        val json = JSONObject(response)
        return TokenSet(
            accessToken = json.getString("access_token"),
            refreshToken = json.optStringOrNull("refresh_token"),
            tokenType = json.optString("token_type", "Bearer"),
            expiresIn = json.optLong("expires_in", 0),
            scope = json.optStringOrNull("scope"),
            idToken = json.optStringOrNull("id_token")
        )
    }

    private fun form(vararg values: Pair<String, String>): String =
        values.joinToString("&") { (key, value) ->
            "${URLEncoder.encode(key, "UTF-8")}=${URLEncoder.encode(value, "UTF-8")}"
        }

    private fun randomUrlSafe(bytes: Int): String {
        val data = ByteArray(bytes)
        SecureRandom().nextBytes(data)
        return java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(data)
    }

    private fun sha256Base64Url(value: String): String =
        java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(
            MessageDigest.getInstance("SHA-256").digest(value.toByteArray(StandardCharsets.US_ASCII))
        )
}

class MaxAuthException(message: String) : Exception(message)

private class SecureTokenStore(private val context: Context, private val clientId: String) {
    private val prefs = context.getSharedPreferences("max_auth_tokens", Context.MODE_PRIVATE)
    private val alias = "max_auth_$clientId"

    fun save(tokens: MaxAuth.TokenSet) {
        val json = JSONObject().apply {
            put("access_token", tokens.accessToken)
            put("refresh_token", tokens.refreshToken)
            put("token_type", tokens.tokenType)
            put("expires_in", tokens.expiresIn)
            put("scope", tokens.scope)
            put("id_token", tokens.idToken)
        }.toString()
        val encrypted = encrypt(json.toByteArray(StandardCharsets.UTF_8))
        prefs.edit().putString("token", Base64.encodeToString(encrypted, Base64.NO_WRAP)).apply()
    }

    fun load(): MaxAuth.TokenSet? {
        val encoded = prefs.getString("token", null) ?: return null
        return runCatching {
            val json = JSONObject(String(decrypt(Base64.decode(encoded, Base64.NO_WRAP)), StandardCharsets.UTF_8))
            MaxAuth.TokenSet(
                json.getString("access_token"),
                json.optStringOrNull("refresh_token"),
                json.optString("token_type", "Bearer"),
                json.optLong("expires_in", 0),
                json.optStringOrNull("scope"),
                json.optStringOrNull("id_token")
            )
        }.getOrNull()
    }

    fun clear() = prefs.edit().remove("token").apply()

    private fun key(): SecretKey {
        val store = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        if (!store.containsAlias(alias)) {
            val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
            generator.init(KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setRandomizedEncryptionRequired(true)
                .build())
            generator.generateKey()
        }
        return (store.getEntry(alias, null) as KeyStore.SecretKeyEntry).secretKey
    }

    private fun encrypt(value: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key())
        return cipher.iv + cipher.doFinal(value)
    }

    private fun decrypt(value: ByteArray): ByteArray {
        val iv = value.copyOfRange(0, 12)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key(), GCMParameterSpec(128, iv))
        return cipher.doFinal(value.copyOfRange(12, value.size))
    }
}

private fun JSONObject.optStringOrNull(name: String): String? =
    if (!has(name) || isNull(name)) null else optString(name).takeIf { it.isNotBlank() }
